Mathlib/Algebra/BigOperators/Associated.lean

```diff
@@ -81,6 +81,35 @@ theorem exists_associated_mem_of_dvd_prod [CancelCommMonoidWithZero α] {p : α}
       exact ⟨q, Multiset.mem_cons.2 (Or.inr hq₁), hq₂⟩
 #align exists_associated_mem_of_dvd_prod exists_associated_mem_of_dvd_prod
 
+open Submonoid in
+/-- Let x, y ∈ α. If x * y can be written as a product of prime elements, then x can be written as
+a product of a unit and prime elements. -/
+theorem Associated_closure_of_mul_mem_closure [CancelCommMonoidWithZero α] :
+    ∀ x y, x * y ∈ closure { r | Prime r } → ∃ z ∈ closure { r : α | Prime r }, x ~ᵤ z := by
+  intros a b hab
+  obtain ⟨m, hm⟩ := exists_multiset_of_mem_closure hab
+  revert a b hm
+  refine' m.induction (p := fun m => ∀ a b, a * b ∈ closure {r | Prime r} →
+    (∃ (_ : ∀ y ∈ m, Prime y), m.prod = a * b) → ∃ z ∈ closure {r | Prime r}, a ~ᵤ z) _ _
+  simp only [Multiset.prod_zero, Multiset.not_mem_zero, Set.mem_setOf_eq, IsEmpty.forall_iff,
+    forall_const, exists_const]
+  exact (fun _ _ _ hprod => ⟨1, one_mem _, associated_one_of_mul_eq_one _ hprod.symm⟩)
+  simp only [Set.mem_setOf_eq, exists_prop, and_imp, Multiset.prod_cons, Multiset.mem_cons,
+    forall_eq_or_imp]
+  intros a s hind x y _ hprime hprime₂ hprod
+  cases' hprime.dvd_mul.1 (Dvd.intro _ hprod) with h₁ h₂
```

Comment:
I have been told that `cases'` would be deprecated soon. The syntax of `cases` is less comfortable but one gets used to it.